### ISSUE TITLE
bug: linked subscription '00000000-0000-0000-0000-000000000000' was not found

### ISSUE THREAD

**Issue by cdaveybjss** [Azure/terraform-azurerm-lz-vending#149]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions

Please paste the output of `terraform version` command from within the initialized directory:

```text
<output of terraform version>
```

Please enter the module version that you are using:

```text
<module version>
```

#### Steps to Reproduce

1. step 1
2. step 2
3. you get it...

#### Additional context

subscription_alias_enabled = true

╷
│ Error: creating/updating "Resource: (ResourceId \"/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxxxxxxxxxx/providers/Microsoft.Network/virtualNetworks/xxxxxxxxxxxx\" / Api Version \"2021-08-01\")": PUT https://management.azure.com/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxxxxxxxxxx/providers/Microsoft.Network/virtualNetworks/xxxxxxxxxxxx
│ --------------------------------------------------------------------------------
│ RESPONSE 403: 403 Forbidden
│ ERROR CODE: LinkedAuthorizationFailed
│ --------------------------------------------------------------------------------
│ {
│   "error": {
│     "code": "LinkedAuthorizationFailed",
│     "message": "The client has permission to perform action 'Microsoft.Network/ddosProtectionPlans/join/action' on scope '/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxxxxxxxxxx/providers/Microsoft.Network/virtualNetworks/xxxxxxxxxxxx', however the linked subscription '00000000-0000-0000-0000-000000000000' was not found. "
│   }
│ }
│ --------------------------------------------------------------------------------
│ 
│ 
│   with module.lz_vending.module.virtualnetwork[0].azapi_resource.vnet["one"],
│   on .terraform/modules/lz_vending/modules/virtualnetwork/main.tf line 37, in resource "azapi_resource" "vnet":
│   37: resource "azapi_resource" "vnet" {
│ 
╵

**Comment by matt-FFFFFF** [comment]
Hi @cdaveybjss 

You seem to be deploying the example in the README, is that right?

If so this includes a hub and spoke peering configuration. In the example the 'hub_network_resource_id' is set to a dummy value.

To get this to work you must supply a real virtual network resource id, or disable hub peering.

Does this help?

**Comment by cdaveybjss** [comment]
Hello,

I am using the example and have the following set to false:

hub_peering_enabled = false
mesh_peering_enabled = false

I am using: vwan_connection_enabled = true

Interestingly, if I use an existing subscription, the vnet is created without error. If I create a subscription with the module (subscription_alias_enabled = true) the error occurs on the first run. If I run the pipeline again with the same variables, the vnet is created.

**Comment by cdaveybjss** [comment]
This is the module configuration:


module "lz_vending" {
  source  = "Azure/lz-vending/azurerm"
  version = "2.0.0"

  # set the default location for resources
  location = var.vending_resources_location

  # subscription variables
  subscription_alias_enabled = true
  subscription_billing_scope = "/providers/Microsoft.Billing/billingAccounts/## REDACTED ##/enrollmentAccounts/## REDACTED ##"
  subscription_display_name  = "sub-${var.vend_id}-${var.environment}-${var.vending_resources_location}-01"
  subscription_alias_name    = "sub-${var.vend_id}-${var.environment}-${var.vending_resources_location}-01"
  subscription_workload      = "Production"

  # use an existing subscription instead
  subscription_id = ""

  # management group association variables
  subscription_management_group_association_enabled = true
  subscription_management_group_id                  = var.vend_management_group_id

  # virtual network variables
  virtual_network_enabled = true
  virtual_networks = {
    vnet1 = {
      resource_group_creation_enabled = true
      resource_group_tags             = var.vending_resources_tags
      resource_group_name             = "rg-vnet-${var.vend_id}-${var.environment}-${var.vending_resources_location}-01"
      name                            = "vnet-${var.vend_id}-${var.environment}-${var.vending_resources_location}-01"
      location                        = var.vending_resources_location
      address_space                   = ["${var.vend_address_space}"]
      hub_peering_enabled             = false
      mesh_peering_enabled            = false

      vwan_connection_enabled = var.vend_virtual_wan_enabled
      vwan_connection_name    = "${var.vend_id}ToHub"
      vwan_hub_resource_id    = "/subscriptions/${var.subscription_id_connectivity}/resourceGroups/rg-vwan-${var.environment}-${var.connectivity_resources_primary_location}-01/providers/Microsoft.Network/virtualHubs/vhub-connectivity-${var.environment}-${var.connectivity_resources_primary_location}-01"
    }
  }

  # role assignments
  role_assignment_enabled = true
  role_assignments = [
    # using role definition name, created at subscription scope
    {
      principal_id   = "## REDACTED ##"
      definition     = "Contributor"
      relative_scope = ""
    }
    # using a custom role definition
    # {
    #   principal_id   = "## REDACTED ##"
    #   definition     = "/providers/Microsoft.Management/MyMg/providers/Microsoft.Authorization/roleDefinitions/ffffffff-ffff-ffff-ffff-ffffffffffff"
    #   relative_scope = ""
    # },
    # using relative scope (to the created or supplied subscription)
    # {
    #   principal_id   = "## REDACTED ##"
    #   definition     = "Owner"
    #   relative_scope = "/resourceGroups/MyRg"
    # },
  ]
}

**Comment by cdaveybjss** [comment]
This is the actual error:

╷
│ Error: creating/updating "Resource: (ResourceId \"/subscriptions/## REDACTED ## /resourceGroups/## REDACTED ##/providers/Microsoft.Network/virtualNetworks/## REDACTED ##\" / Api Version \"2021-08-01\")": PUT https://management.azure.com/subscriptions/## REDACTED ##/resourceGroups/## REDACTED ##/providers/Microsoft.Network/virtualNetworks/## REDACTED ##
│ --------------------------------------------------------------------------------
│ RESPONSE 403: 403 Forbidden
│ ERROR CODE: LinkedAuthorizationFailed
│ --------------------------------------------------------------------------------
│ {
│   "error": {
│     "code": "LinkedAuthorizationFailed",
│     "message": "The client has permission to perform action 'Microsoft.Network/ddosProtectionPlans/join/action' on scope '/subscriptions/## REDACTED ##/resourceGroups/## REDACTED ##/providers/Microsoft.Network/virtualNetworks/## REDACTED ##', however the linked subscription '00000000-0000-0000-0000-000000000000' was not found. "
│   }
│ }
│ --------------------------------------------------------------------------------
│
│
│   with module.lz_vending.module.virtualnetwork[0].azapi_resource.vnet["one"],
│   on .terraform/modules/lz_vending/modules/virtualnetwork/main.tf line 37, in resource "azapi_resource" "vnet":
│   37: resource "azapi_resource" "vnet" {
│
╵

**Comment by matt-FFFFFF** [comment]
> │ "error": { │ "code": "LinkedAuthorizationFailed", │ "message": "The client has permission to perform action 'Microsoft.Network/ddosProtectionPlans/join/action' on scope '/subscriptions/## REDACTED ##/resourceGroups/## REDACTED ##/providers/Microsoft.Network/virtualNetworks/## REDACTED ##', however the linked subscription '00000000-0000-0000-0000-000000000000' was not found. " │ } │ }

Hi,

This error is for linking a ddos protection plan, in fact it is on our backlog at #125 😄 

Can you have a look at the plan output to see if this module is somehow trying to link the vnet to a ddos protection plan?

Else, is there anything in your environment that could be causing this (policy, etc)?

**Comment by cdaveybjss** [comment]
There is nothing in the plan to indicate a DDoS protection plan nor can I find any policy.

I have also tried deploying to a newly created management group that contains no policies with the same result.

The error would seem to indicate that the account being used does in fact have permission to perform the action but isn’t passing the correct subscription id.

I don’t understand why there is a DDoS protection plan being referenced.

**Comment by cdaveybjss** [comment]
If I run the same pipeline with the same values again, the deployment is successful. We don’t want to run the pipeline twice.

If I run the deployment with an existing subscription, the deployment is successful. We don’t want to use existing subscriptions.

**Comment by matt-FFFFFF** [comment]
Hi @cdaveybjss 

Thanks for following up. Some questions:

Is there a correlation ID in the error? I can use this to look through the ARM logs to try and see why this was failing. If you can supply more details over email then I'm happy to help. Send me the subscription IDs to matt(dot)white at...

Can you tell me what resource in the plan generated the error?

Also can you tell me if you have a firewall in the vWAN hub, is so does that firewall network have a DDOS plan [associated](https://learn.microsoft.com/azure/firewall-manager/configure-ddos#associate-a-ddos-protection-plan)?

**Comment by matt-FFFFFF** [comment]
(internal only) see AB#26298

**Comment by matt-FFFFFF** [comment]
Hey @cdaveybjss 

This feels like it's the result of a modify policy included in ALZ.

Please can you run the following az cli command, replacing the subscription ID with the newly created one:

```shell
az policy assignment list --disable-scope-strict-match --scope /subscriptions/0000..
```

I'm specifically looking for this policy definition reference id:

https://www.azadvertizer.net/azpolicyadvertizer/94de2ad3-e0c1-4caf-ad78-5d47bbc83d3d.html

This is a modify policy and would fit with the behaviour I think.

**Comment by cdaveybjss** [comment]
Apologies for the delay in response.

This has now been resolved. Your suggestion was most helpful in finding the solution.

There are no DDoS plans deployed by the counterpart caf module, but someone had defined a policy and assigned it by way of archetype. After inspecting the output from the deployment and running the policy command, I could see the reference and saw one was being created and assigned during deployment. There is no bug in the vending module at all.

The policy was defined with a parameter value of /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/etc. as a place holder for when a DDoS plan might be used in the future. Since this policy won’t work and doesn’t hold any value for now, I removed the definition and assignment. Vending a new subscription and VNet is now successful.
